test/metatx/ERC2771Forwarder.test.js

```diff
@@ -208,21 +216,30 @@ contract('ERC2771Forwarder', function (accounts) {
     });
 
     it('bubbles out of gas', async function () {
-      const receiver = await CallReceiverMock.new();
-      const gasAvailable = 100000;
-      this.requestData.to = receiver.address;
-      this.requestData.data = receiver.contract.methods.mockFunctionOutOfGas().encodeABI();
-      this.requestData.gas = 1000000;
-
+      this.requestData.data = this.receiver.contract.methods.mockFunctionOutOfGas().encodeABI();
+      this.requestData.gas = 1_000_000;
       this.requestData.signature = this.sign(this.alice.getPrivateKey());
 
+      const gasAvailable = 100_000;
       await expectRevert.assertion(this.forwarder.execute(this.requestData, { gas: gasAvailable }));
 
       const { transactions } = await web3.eth.getBlock('latest');
       const { gasUsed } = await web3.eth.getTransactionReceipt(transactions[0]);
 
       expect(gasUsed).to.be.equal(gasAvailable);
     });
+
+    it('bubbles out of gas forced by the relayer', async function () {
+      // A malicious relayer would provide a gas limit that is too low to execute the request.
+      const estimate = await this.estimateRequest(this.request);
+      const gasAvailable = estimate + 2_000; // Add some gas to let the first-level transaction pass
+      await expectRevert.outOfGas(this.forwarder.execute(this.requestData, { gas: gasAvailable }));
```

Comment:
My understanding of the test is
- the request is the normal one (`mockFunction()` with `100_000` gas)
- you estimate how much is actually needed (not 100_000)
- you give slightly more (estimate + 2000)
  - the forwarder use some gas for "pre-processing"
  - the forwarder passes 63/64 which you expect to be less than `estimate`
  - the inner call should revert (out of gas)
  - the forwarder call should detect that and consume all the gas (even though some is left)
    - that is what is checked by `expect(gasUsed).to.be.equal(gasAvailable);`
    
    
I took me a while to understand all that. Am I correct ? IMO we should add some documentation to explain how the test works.

Reply:
Yes, all of that is correct, and now that you mention it I agree it may be difficult for others to parse. I added comments for both tests (single and batch).